examples/qualcomm/llama2/qaihub_runner/runner.cpp

```diff
@@ -0,0 +1,357 @@
+#include <executorch/runtime/core/exec_aten/util/scalar_type_util.h>
+#include <executorch/runtime/platform/log.h>
+
+#include "arm_neon.h"
+
+namespace torch {
+namespace executor {
+
+namespace {
+static constexpr auto kTopp = 0.9f;
+void printReport(const Runner::Stats& stats);
+std::string statsToJsonString(const Runner::Stats& stats);
+} // namespace
+
+Runner::Runner(
+    const std::vector<std::string>& models_path,
+    const std::vector<std::string>& pos_embs_path,
+    const std::string& tokenizer_path,
+    const int eval_mode,
+    const float temperature,
+    const float logits_scale,
+    const int logits_offset)
+    : tokenizer_path_(tokenizer_path),
+      temperature_(temperature),
+      bos_id_(1),
+      eos_id_(2),
+      n_bos_(1),
+      n_eos_(1),
+      vocab_size_(32000),
+      max_seq_len_(1024),
+      eval_mode_(eval_mode),
+      stats_({}),
+      logits_scale_(logits_scale),
+      logits_offset_(logits_offset) {
+  for (size_t i = 0; i < models_path.size(); ++i) {
+    modules_.push_back(std::make_shared<Module>(
+        models_path[i], Module::LoadMode::MmapUseMlockIgnoreErrors));
+    ET_LOG(Info, "creating module: model_path=%s", models_path[i].c_str());
+  }
+  ET_LOG(Info, "creating runner: tokenizer_path=%s", tokenizer_path_.c_str());
+  switch (eval_mode_) {
+    case EvalMode::kBert:
+      io_mem_ = std::make_unique<BertMemory>(pos_embs_path, modules_);
+      break;
+    case EvalMode::kKVCached:
+      io_mem_ = std::make_unique<KVCachedMemory>(pos_embs_path, modules_);
+      break;
+    default:
+      ET_CHECK_MSG(false, "unsupported evaluation mode");
+  }
+  ET_LOG(Info, "creating io_memory");
+}
+
+bool Runner::is_loaded() const {
+  bool loaded = true;
+  for (const std::shared_ptr<Module>& module : modules_) {
+    loaded &= module->is_loaded();
+  }
+  return loaded && tokenizer_ && sampler_;
+}
+
+Error Runner::load() {
+  if (is_loaded()) {
+    return Error::Ok;
+  }
+  for (std::shared_ptr<Module>& module : modules_) {
+    ET_CHECK_OK_OR_RETURN_ERROR(module->load_method("forward"));
+  }
+
+  // load tokenizer
+  tokenizer_ = std::make_unique<BPETokenizer>();
+  tokenizer_->load(tokenizer_path_);
+
+  // create sampler
+  sampler_ = std::make_unique<Sampler>(
+      vocab_size_,
+      temperature_,
+      kTopp,
+      static_cast<unsigned long long>(std::time(nullptr)));
+
+  // prepare io
+  auto methods_meta = get_methods_meta();
+  io_mem_->prepare_io(methods_meta);
+  return Error::Ok;
+}
+
+int32_t Runner::logitsToToken(const Tensor& logits_tensor) {
+  static std::vector<float> logits_f(vocab_size_);
+  static int32x4_t offset = vmovq_n_s32(logits_offset_);
+  static float32x4_t scale = vmovq_n_f32(logits_scale_);
+  const uint16_t* logits = logits_tensor.data_ptr<uint16_t>();
+  // dequantize
+  for (int i = 0; i < vocab_size_; i += 4) {
+    const uint16_t* in = logits + i;
+    float* out = logits_f.data() + i;
+    int32x4_t q = {in[0], in[1], in[2], in[3]};
```

Comment:
I would recommend issuing a vectorized load and then vcvtq_s32_u16 here because this probably does a bunch of slow fmovs, though I haven't checked generated assembly

Reply:
Thanks for this, I think there has no `vcvtq_u16_s32` intrinsic. Change code a little bit to omit `fmov` (verified with compiler explorer).